solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -377,8 +377,72 @@ private StatefulCollectionWatch compute(
     }
   }
 
-  private static class StatefulCollectionWatch extends CollectionWatch<DocCollectionWatcher> {
+  private class StatefulCollectionWatch extends CollectionWatch<DocCollectionWatcher> {
     private DocCollection currentState;
+
+    Watcher persistentWatcher;
+
+    final String coll;
+    private final String collectionPath;
+
+    private StatefulCollectionWatch(String coll) {
+      this.coll = coll;
+      this.collectionPath = DocCollection.getCollectionPath(coll);
+    }
+
+    public void handleWatch(WatchedEvent event) {
+      if (!collectionWatches.watchedCollections().contains(coll)) {
+        removePersistentWatch();
+        return;
+      }
+      log.debug("an event happened for {}, event: {}", coll, event.toString());
+      DocCollection collectionState = getCollection(coll);
+      if (collectionPath.equals(event.getPath())) {
+        DocCollection newState = null;
+        try {
+          newState = fetchCollectionState(coll, null);
+        } catch (KeeperException.SessionExpiredException
+            | KeeperException.ConnectionLossException e) {
+          log.warn("ZooKeeper watch triggered, but Solr cannot talk to ZK: ", e);
+        } catch (KeeperException e) {
+          log.error("exception for collection: [{}]", coll, e);
+          throw new ZooKeeperException(ErrorCode.SERVER_ERROR, "A ZK error has occurred", e);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          log.error("exception for collection: [{}]", coll, e);
+        }
+        collectionWatches.updateDocCollection(coll, newState);
+        synchronized (getUpdateLock()) {
+          constructState(Collections.singleton(coll));
+        }
+      }
+      if (collectionState!= null && collectionState.isPerReplicaState()) {
+        String path = event.getPath();
+        if ((event.getType() == EventType.NodeCreated || event.getType() == EventType.NodeDeleted)
+            && path.length() > collectionState.getZNode().length()
+            && path.startsWith(path)) {
+          collectionState =
+              collectionState.copyWith(
+                  PerReplicaStatesFetcher.fetch(collectionPath, zkClient, null));
```

Comment:
Here path represents the prs state `/collections/perReplicaState_test/state.json/core_node2:1:A`. It means we don't need to fetch all the children, we can just update the `core_node2` replica state. 

Similarly, we need to create new state in zk, and then need to delete old state. Which I think already happening, but these create and delete event needs to handle appropriately.

Reply:
I know . I went with that approach first. But, there were some errors. I'll need to fix them before that optimisation can be brought in

Reply:
I'm curious what you think it might take to do that. I think a lot of the value here could be minimizing what we have to read from Zookeeper and what gets updated in state.

I also think that'll help show the full value of the persistent watches for any upstream PR/conversation.

Reply:
let's say 



1.  the node has a `PRS` object with `cversion = n`
2. at `T1` the node gets an event `CREATE (core_node4:2:A)` and `cversion = n+1` . Please note that the event doesn't send the stat of the node at all.

now we create a new `PRS` object and using the delta of `CREATE (core_node4:2:A)` . To create the new PRS object , we need the new `cversion` . We need to do a `zkClient.exists()` call  which happens at `T2`.  `T2 > T1` and may be there are other events that would have happened between `T1` and `T2` and the `cversion` would have gotten bumped up to `n+2` or `n+3.` Throughout the code we compare the `cversion` of the `PRS` objects to check if it is the latest or not. This will lead to inconsistencies

Reply:
@noblepaul how about storing the `creation time` for that replica update.  Then get `modified time` of parent node (state.json) file. if its changed then fetch it again.

I'm assuming when child nodes are created, parent node's `modified time` changes. Maybe worth to understand this property or other property of parent node.

Reply:
>I'm assuming when child nodes are created, parent node's modified time changes.

Well, no. The parent node `mtime` changes when you perform a `set` data on `state.json`

@hiteshk25 I'll try using the `ctime` of the node `core_node4:2:A` as the version instead of `cversion`. 

So, if another replica `core_nodeX:2:A` is created after `core_node4:2:A`, it will have a newer timestamp. so comparison will always give correct value

The problem is what happens when we read all values at once. we do not know which node was created the last. We will need to just use the current timestamp of the invoking client as the `ctime`

Reply:
Sorry trying to catch up and agree with everything said so far! 😊 

From my understanding of the fundamental design of watcher pattern (old watch or new recursive permanent watch) appears to be:
1. Receive a notification on path, with event type/state, no stats nor data here
2. Fetch the data on the path, get the actual data and stats (versions)

However, how we are trying to use it:
1. Receive a notification on path and try to parse the path from the notification and use it as data all in one go (but missing version)
2. In order to get a version (that corresponds to all children), we either try to fetch the stats (instead of full data), whether from the `state.json`'s `cVersion` or from the actual PRS entry `ctime` as discussed later on.

This seems to be a tricky problem as data and version no longer comes from the same call, and such gap could trigger race condition as @noblepaul described for `cVersion` while using `ctime`/timestamps might have other concerns (only millisec precision/collision, time drift, client time vs zk time etc). Also not exactly sure if `ctime` would work in `PerReplicaStates#fetch`, which it currently gets and checks on the `Stat.cversion()` on the state.json.

Wondering if it's possible to use the `sequence=true` flag when creating the path and use that as our `cversion`? With some simply test the child node generated appears to have an increasing counter shared across all siblings? that way when we get the notification we will know the data (the actual first half of the path) and the version (sequence number) 🤔

Reply:
You are right @patsonluk 
My idea of using `ctime` of the node wouldn't  work. 

`PerReplicaStatesFetcher#fetch()` needs to know if the current state in ZK is newer than the `PerReplicaStatesFetcher` object we have locally. That is not possible at all

 I shall try you suggestion of using `sequence=true`

Reply:
> Then we might be able to get the notification with both data (the actual first half of the path) and the cversion (sequence number) in a single step? 🤔
 
 We may be able to get the version while writing. But the real challenge is to check if a given PRS object has the latest data. we may need to read all the children and parse their names to get the highest sequence number

Reply:
I think it would also be reasonable to fetch the data for the individual PRS path that changed so it doesn't necessarily need to be a single step or notifying the watch and getting data all at once, but we should avoid having to fetch all PRS states when a single one changes.

Reply:
>I think it would also be reasonable to fetch the data for the individual PRS path that changed

That is the easy part.

But it's not enough. If we have to decide whether the PRS object we have is stale, there is no way to do that. We have to read all the child nodes and find which one is the newest. That is very inefficient